src/FilterList.js

```diff
@@ -2,14 +2,27 @@ import React from 'react'
 import _ from 'lodash/fp'
 import F from 'futil-js'
 import { observer, inject } from 'mobx-react'
-import { Flex, Dynamic, Popover } from './layout'
+import { Flex, Dynamic, Popover, Modal, NestedPicker } from './layout'
+import { fieldsToOptions } from './FilterAdder'
 import { withStateLens } from './utils/mobx-react-utils'
 import InjectTreeNode from './utils/injectTreeNode'
 import DefaultIcon from './DefaultIcon'
 import { bdJoin } from './styles/generic'
 import { newNodeFromType } from './utils/search'
 
-export let FilterActions = ({ node, tree, fields, Item }) => (
+let FieldPicker = ({ options, isOpen, onChange }) => (
```

Comment:
You can save yourself moving state around (`isOpen` in this case) if you move the button here and use `withState` here as well.

Reply:
The button needs to be separate because the modal has to be detached from the button (so that it's not rendered inside the popover)

Reply:
Don't we use a portal for the modal? Why does it render inside the popover?

Reply:
It's not rendered inside the popover, but it is killed when the popover closes, which (if we did put the modal inside) would happen as soon as you try to interact with the modal, or even as soon as you click the item to open the modal in the first place.

It's possible (in theory) to hack around the normal popover functionality to avoid these problems, but it's not worth it - best to just keep the modal outside.

Reply:
Yeah even with portals if the parent unmounts so does the child (Modal in this case).

> It's possible (in theory) to hack around the normal popover functionality

I wouldn't call our current popover "normal" lol. The name on this https://github.com/airbnb/react-outside-click-handler is misleading. It should just handle outside clicks but it closes the popover when it's clicked inside :(

Reply:
I would add a comment somewhere though, because someone could have the same question about why isn't the state closer to where it's used.